### ISSUE TITLE
docker-compose单机部署，运行任务报错 message info: dispatch task to party failed, detail: Failed to get all dataset's meta, no handler triggered. expected dataset size: 3, but get: 1

### ISSUE THREAD

**Issue by Klein920116** [primihub/primihub#782]
./primihub-cli --task_config_file="example/FL/linear_regression/hfl_dpsgd.json"
I20240604 10:54:44.179337    97 cli.cc:570] use tls: 0
I20240604 10:54:44.180269    97 cli.cc:580] SDK SubmitTask to: 127.0.0.1:50050
W20240604 10:54:44.183193    97 cli.cc:208] task_manager is not one of party in this task, ignore....
I20240604 10:54:44.189026    97 cli.cc:351] Task Request: task { name: "demoTask" params { param_map { key: "component_params" value { var_type: STRING value_string: "{\"common_params\":{\"alpha\":0.0001,\"batch_size\":100,\"delta\":0.001,\"global_epoch\":10,\"id\":\"id\",\"l2_norm_clip\":1.0,\"label\":\"y\",\"learning_rate\":0.1,\"local_epoch\":1,\"method\":\"DPSGD\",\"model\":\"HFL_linear_regression\",\"noise_multiplier\":2.0,\"print_metrics\":true,\"process\":\"train\",\"secure_mode\":true,\"selected_column\":null,\"task_name\":\"HFL_linear_regression_dpsgd_train\"},\"role_params\":{\"Alice\":{\"data_set\":\"fl_fake_data\",\"metric_path\":\"data/result/server_metrics.json\"},\"Bob\":{\"data_set\":\"regression_hfl_train_client1\",\"metric_path\":\"data/result/Bob_metrics.json\",\"model_path\":\"data/result/Bob_model.pkl\"},\"Charlie\":{\"data_set\":\"regression_hfl_train_client2\",\"metric_path\":\"data/result/Charlie_metrics.json\",\"model_path\":\"data/result/Charlie_model.pkl\"}},\"roles\":{\"client\":[\"Bob\",\"Charlie\"],\"server\":\"Alice\"}}" } } } task_info { task_id: "200" job_id: "100" request_id: "9edc152f-e564-4e13-8633-7a5ac067f95b" } party_datasets { key: "Alice" value { data { key: "data_set" value: "fl_fake_data" } } } party_datasets { key: "Bob" value { data { key: "data_set" value: "regression_hfl_train_client1" } } } party_datasets { key: "Charlie" value { data { key: "data_set" value: "regression_hfl_train_client2" } } } } 
I20240604 10:54:44.189066    97 cli.cc:74] no download cfg, ignore...
I20240604 10:54:44.189126    97 cli.cc:456] SubmitTask: RequestId: 9edc152f-e564-4e13-8633-7a5ac067f95b 
E20240604 10:54:44.204960    97 cli.cc:466] party count from reply is: 0
 message info: dispatch task to party failed, detail: Failed to get all dataset's meta, no handler triggered. expected dataset size: 3, but get: 1
E20240604 10:54:44.205252    97 cli.cc:609] SubmitTask failed

**Comment by phoenix20162016** [comment]
https://github.com/primihub/primihub-deploy/releases/tag/1.6.11 用部署脚本部署

**Comment by phoenix20162016** [comment]
可以源码自己编译，物理机启动node节点进行体验 @Klein920116

**Comment by Klein920116** [comment]
> 可以源码自己编译，物理机启动node节点进行体验 @Klein920116

服务器单机部署不可以吗？

**Comment by phoenix20162016** [comment]
暂时这个是有点问题的

**Comment by Klein920116** [comment]
> 暂时这个是有点问题的

那体验的话只能是在多台物理机都进行编译了？

**Comment by Klein920116** [comment]
> 暂时这个是有点问题的

可以告诉一下你们的文档里哪些是可以正常跑的吗？多机/异地部署可以吗？快速启动也有问题是吗？

**Comment by phoenix20162016** [comment]
不是的，现在部署脚本有两种模式，单机部署和多机部署，你的这种情况可以用单机部署模式

**Comment by phoenix20162016** [comment]
https://github.com/primihub/primihub-deploy/tree/main/docker-all-in-one

**Comment by Klein920116** [comment]
> ./primihub-cli --task_config_file="example/FL/linear_regression/hfl_dpsgd.json" I20240604 10:54:44.179337 97 cli.cc:570] use tls: 0 I20240604 10:54:44.180269 97 cli.cc:580] SDK SubmitTask to: 127.0.0.1:50050 W20240604 10:54:44.183193 97 cli.cc:208] task_manager is not one of party in this task, ignore.... I20240604 10:54:44.189026 97 cli.cc:351] Task Request: task { name: "demoTask" params { param_map { key: "component_params" value { var_type: STRING value_string: "{"common_params":{"alpha":0.0001,"batch_size":100,"delta":0.001,"global_epoch":10,"id":"id","l2_norm_clip":1.0,"label":"y","learning_rate":0.1,"local_epoch":1,"method":"DPSGD","model":"HFL_linear_regression","noise_multiplier":2.0,"print_metrics":true,"process":"train","secure_mode":true,"selected_column":null,"task_name":"HFL_linear_regression_dpsgd_train"},"role_params":{"Alice":{"data_set":"fl_fake_data","metric_path":"data/result/server_metrics.json"},"Bob":{"data_set":"regression_hfl_train_client1","metric_path":"data/result/Bob_metrics.json","model_path":"data/result/Bob_model.pkl"},"Charlie":{"data_set":"regression_hfl_train_client2","metric_path":"data/result/Charlie_metrics.json","model_path":"data/result/Charlie_model.pkl"}},"roles":{"client":["Bob","Charlie"],"server":"Alice"}}" } } } task_info { task_id: "200" job_id: "100" request_id: "9edc152f-e564-4e13-8633-7a5ac067f95b" } party_datasets { key: "Alice" value { data { key: "data_set" value: "fl_fake_data" } } } party_datasets { key: "Bob" value { data { key: "data_set" value: "regression_hfl_train_client1" } } } party_datasets { key: "Charlie" value { data { key: "data_set" value: "regression_hfl_train_client2" } } } } I20240604 10:54:44.189066 97 cli.cc:74] no download cfg, ignore... I20240604 10:54:44.189126 97 cli.cc:456] SubmitTask: RequestId: 9edc152f-e564-4e13-8633-7a5ac067f95b E20240604 10:54:44.204960 97 cli.cc:466] party count from reply is: 0 message info: dispatch task to party failed, detail: Failed to get all dataset's meta, no handler triggered. expected dataset size: 3, but get: 1 E20240604 10:54:44.205252 97 cli.cc:609] SubmitTask failed

我出现这个问题就是用单机部署。

**Comment by phoenix20162016** [comment]
docker-compose down
然后再执行bash deploy.sh

**Comment by Klein920116** [comment]
> 可以源码自己编译，物理机启动node节点进行体验 @Klein920116

那您的这个回答是解决哪个问题呢?

**Comment by Lightzeng** [comment]
遇到了同样的问题，文档上说注册自定义数据 ./bazel-bin/src/primihub/reg-cli --config_file=./example/register_example/reg_dataset_meta.json，没有找到reg-cli这个

最新的注册接口可以用primihub-cli --type=reg --task_config_file=xxxx
用cli注册的数据在页面端是不可见的 @Lightzeng

**Comment by phoenix20162016** [comment]
> > 可以源码自己编译，物理机启动node节点进行体验 @Klein920116
> 
> 那您的这个回答是解决哪个问题呢?
开发一般用这种模式，我现在暂时用这个模式还没遇到什么问题，可以直接用cli进行体验

**Comment by Klein920116** [comment]
> docker-compose down 然后再执行bash deploy.sh

我刚刚照着做了，还是出现问题。需要用1.6.11的代码？

**Comment by phoenix20162016** [comment]
https://github.com/primihub/primihub-deploy/releases/tag/1.6.11-fix 这个是一个修复版本，还是比较稳定的， 还有你可以自己注册数据集，

**Comment by phoenix20162016** [comment]
> 遇到了同样的问题，文档上说注册自定义数据 ./bazel-bin/src/primihub/reg-cli --config_file=./example/register_example/reg_dataset_meta.json，没有找到reg-cli这个

注册数据集暂时从前端吧

**Comment by Klein920116** [comment]
> https://github.com/primihub/primihub-deploy/releases/tag/1.6.11-fix 这个是一个修复版本，还是比较稳定的， 还有你可以自己注册数据集，

**1.6.11-fix版本**
./primihub-cli --task_config_file="example/FL/linear_regression/hfl_plaintext.json"
I20240604 14:58:30.369902    33 cli.cc:621] use tls: 0
I20240604 14:58:30.370611    33 cli.cc:632] SDK SubmitTask to: 127.0.0.1:50050
W20240604 14:58:30.373872    33 cli.cc:308] task_manager is not one of party in this task, ignore....
I20240604 14:58:30.375000    33 cli.cc:510]  SubmitTask...
E20240604 14:58:30.464391    33 cli.cc:519] party count from reply is: 0
I20240604 14:58:30.464946    33 cli.cc:645] SubmitTask time cost(ms): 91

**Comment by Klein920116** [comment]
> https://github.com/primihub/primihub-deploy/releases/tag/1.6.11-fix 这个是一个修复版本，还是比较稳定的， 还有你可以自己注册数据集，

**1.6.11-fix一样不行啊**
./primihub-cli --task_config_file="example/FL/xgboost/hetero_xgb.json"
I20240604 15:13:30.008509    33 cli.cc:621] use tls: 0
I20240604 15:13:30.009887    33 cli.cc:632] SDK SubmitTask to: 127.0.0.1:50050
W20240604 15:13:30.017856    33 cli.cc:308] task_manager is not one of party in this task, ignore....
I20240604 15:13:30.019464    33 cli.cc:510]  SubmitTask...
E20240604 15:13:30.075982    33 cli.cc:519] party count from reply is: 0
I20240604 15:13:30.077428    33 cli.cc:645] SubmitTask time cost(ms): 62

**Comment by Klein920116** [comment]
docker logs -f primihub-node0
I20240604 15:12:51.354665     1 service.cc:205] 💾 Restore dataset from local storage...
I20240604 15:12:52.529559     1 service.cc:171] 📃 Load default datasets from config: /app/config/primihub_node0.yaml
I20240604 15:12:54.054924    22 node_impl.cc:98] start ManageTaskThread
I20240604 15:12:54.054999     1 main.cc:70] server runing in no tls mode
I20240604 15:12:54.055938     1 main.cc:106]  💻 Node listening on port: 50050
I20240604 15:13:30.023133    24 node_impl.cc:173] start to schedule task, task_type: 0
I20240604 15:13:30.023185    24 node_impl.cc:133] Start create worker node0 worker id: dc475852-0bff-4647-9bd0-7dc3bc0a6de7
I20240604 15:13:30.023226    24 node_impl.cc:139] Fininsh create worker node0 worker id: dc475852-0bff-4647-9bd0-7dc3bc0a6de7
I20240604 15:13:30.023257    24 node_impl.cc:182] job_id: 100, task_id: 200, sub_task_id: , request_id: dc475852-0bff-4647-9bd0-7dc3bc0a6de7, insert worker id: dc475852-0bff-4647-9bd0-7dc3bc0a6de7 into task_scheduler_map success
I20240604 15:13:30.023434    24 py_parser.cc:29] party_datasets: 2
I20240604 15:13:30.023473    24 py_parser.cc:43] party_name: Charlie dataset_id: train_hetero_xgb_guest
I20240604 15:13:30.023501    24 py_parser.cc:43] party_name: Bob dataset_id: train_hetero_xgb_host
I20240604 15:13:30.023542    24 parser.cc:126] Finding meta list from datasets...
E20240604 15:13:30.074502    24 grpc_impl.cc:203] Failed to get all dataset's meta, no handler triggered.datasets_with_tag size: 2 
E20240604 15:13:30.074587    24 parser.cc:173] no party access config found for dispatch
E20240604 15:13:30.074606    24 node_impl.cc:204] dispatch task to party failed

**Comment by phoenix20162016** [comment]
首先各个节点需要先建立连接，通过页面操作，这样才能保证注册的数据集在各个参与方之间同步

**Comment by Klein920116** [comment]
> 首先各个节点需要先建立连接，通过页面操作，这样才能保证注册的数据集在各个参与方之间同步

单机部署后还需要各个节点建立连接吗？

**Comment by phoenix20162016** [comment]
是的，需要建立连接

**Comment by Klein920116** [comment]
> 是的，需要建立连接

请问你们有单独的算法模块可以直接被调用的吗？

**Comment by phoenix20162016** [comment]
https://github.com/primihub/fedps

**Comment by phoenix20162016** [comment]
> > 是的，需要建立连接
> 
> 请问你们有单独的算法模块可以直接被调用的吗？

通过源码编译，bash start_server.sh 启动sever，然后通过cli可以执行任务

**Comment by Klein920116** [comment]
> 然后通过cli可以执行任务

有没有什么方式可以让我们快速地跑一个联邦学习任务？

**Comment by phoenix20162016** [comment]
明天吧，我现在正在修复用repo种docker-compose启动的问题

**Comment by phoenix20162016** [comment]
如果你是x86架构的服务器的话，现在可以在primihub repo中使用docker-compose up -d 启动服务，然后登录primihub-node0 发起任务就可以了
